docs/arcadia-users-group/20231104-testing-concepts/lesson.md

```diff
@@ -0,0 +1,572 @@
+
+## Continuous integration
+
+Continuous integration makes running tests as easy as possible by integrating the test suite into the development process. 
+Every time a change is made to the repository, the continuous integration system builds and checks that code.
+
+Based on instructions you provide, a continuous integration server can:
+
+- check out new code from a repository
+- spin up instances of supported operating systems (i.e. various versions of OSX, Linux, Windows, etc.).
+- spin up those instances with different software versions (i.e. python 2.7 and python 3.0)
+- run the build and test scripts
+- check for errors
+- report the results.
+
+Since the first step the server conducts is to check out the code from a repository, we'll need to put our code online to make use of this kind of server.
+
+### Set Up a Mean Git Repository on GitHub
+
+Our `mean.py` `test_mean.py` files can be the contents of a repository on GitHub.
+
+* Go to GitHub and [create a repository](https://github.com/new) called aug-mean. Do this in your own user account and don't add any files to the new repo (no README/LICENSE, etc.).
+* Turn the `aug-mean` directory that we've been working in on your computer into a git repository following the "…or create a new repository on the command line" instructions on GitHub:
+```
+echo "# aug-mean" >> README.md
+git init
+git add README.md
+git commit -m "first commit"
+git branch -M main
+git remote add origin git@github.com:yourusername/aug-mean.git
+git push -u origin main
+```
+* Create a new branch (`git checkout -b yourinitials/init`).
+* Use git to `add`, `commit`, and `push` the two files `mean.py` and `test_mean.py` to GitHub.
+
+### GitHub Actions
+
+[GitHub Actions](https://github.com/features/actions) is a continuous integration service provided by GitHub.
+It's integrated directly into GitHub repositories and does not require additional accounts or external services.
+
+To use GitHub Actions, create a directory called `.github` and within it, create another directory called `workflows`.
+
+```
+mkdir -p .github/workflows
+```
+
+Inside the `workflows` directory, you can create a YAML file (e.g. `ci.yml`) to define your continuous integration process:
+
+```
+name: Continuous Integration
+
+on: [push, pull_request]
+
+jobs:
+  build:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macos-latest, windows-latest]
+        python-version: [3.7, 3.8, 3.9, '3.10', '3.11'] 
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r requirements.txt
+    - name: Run tests
+      run: pytest
```

Comment:
given that this is an intro to testing, and also that github actions are a bit opaque to begin with, it might be worth using a simpler action here without the matrix stuff, e.g. 

```
jobs:
  build:
    runs-on: ubuntu-latest 
    steps:
    - uses: actions/checkout@v2
    - name: Set up Python 3.10 
      uses: actions/setup-python@v2
      with:
        python-version: 3.10
    - name: Install dependencies
      run: |
        python -m pip install --upgrade pip
        pip install -r requirements.txt
    - name: Run tests
      run: pytest
```

Reply:
I agree with this. Or explain to them what all the matrix stuff does. 

We should also explain at the end that while this specific example uses `requirements.txt` to install dependencies, it can technically be installed in other ways as well (including conda)